src/TorchSharp/Tensor/Storage.cs

```diff
@@ -0,0 +1,531 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+
+namespace TorchSharp
+{
+    public static partial class torch
+    {
+        /// <summary>
+        /// A torch.Storage is a contiguous, one-dimensional array of a single data type.
+        /// Every tensor has a corresponding storage of the same data type.
+        /// </summary>
+        public abstract class Storage
+        {
+            protected Storage()
+            {
+            }
+
+            protected Storage(Tensor tensor)
+            {
+                _tensor = tensor;
+                data_ptr();
+            }
+
+            protected Storage(Tensor tensor, IntPtr data_ptr)
+            {
+                _tensor = tensor;
+                _tensor_data_ptr = (data_ptr == IntPtr.Zero) ? this.data_ptr() : data_ptr;
+            }
+
+            internal static Storage<T> Create<T>(Tensor tensor) where T : unmanaged
+            {
+                var type = typeof(T);
```

Comment:
I forgot to mention one thing here, if you want to have better optimization for this method, try to change the `switch` statement into `if` statements. The JIT compiler optimization will kick off better when using  `if`.